breakpad-symbols/src/sym_file/mod.rs

```diff
@@ -69,267 +69,182 @@ pub mod walker;
-            let input = buf.data();
-            let consumed = parser.parse_more(input)?;
-            total_consumed += consumed as u64;
-
-            // Give the other consumer of this Reader a chance to use this data.
-            callback(&input[..consumed]);
+            let input = self.buf.data();
+            let consumed = self.parser.parse_more(input)?;
+            self.total_consumed += consumed as u64;
 
             // Remember for the next iteration if all the input was consumed.
-            fully_consumed = input.len() == consumed;
-            buf.consume(consumed);
+            self.fully_consumed = input.len() == consumed;
+            self.buf.consume(consumed);
         }
     }
 
+    pub fn finish(self) -> Result<SymbolFile, SymbolError> {
+        if self.total_consumed == 0 {
+            // We never consumed any bytes, so this is an empty file?
+            // Give a better error message for that.
+            Err(SymbolError::ParseError(
+                "empty SymbolFile (probably something wrong with your debuginfo tooling?)",
+                0,
+            ))
+        } else if self.fully_consumed {
+            // Success! The last iteration cleared the buffer and we still got
+            // no more bytes, so that's a proper EOF with a complete parse!
+            Ok(self.parser.finish())
+        } else {
+            // Ok give up, this input is just impossible.
+            Err(SymbolError::ParseError(
+                "unexpected EOF during parsing of SymbolFile (or a line was too long?)",
+                self.parser.lines,
+            ))
+        }
+    }
+}
+
+impl SymbolFile {
     // Parse a SymbolFile from bytes.
     pub fn from_bytes(bytes: &[u8]) -> Result<SymbolFile, SymbolError> {
-        Self::parse(bytes, |_| ())
+        let mut parser = StreamingSymbolParser::new();
+        parser.add_input(bytes)?;
+        parser.finish()
     }
 
     // Parse a SymbolFile from a file.
     pub fn from_file(path: &Path) -> Result<SymbolFile, SymbolError> {
-        let file = File::open(path)?;
-        Self::parse(file, |_| ())
+        let mut file = File::open(path)?;
+        let mut buf = vec![0; 1024 * 8];
+        let mut parser = StreamingSymbolParser::new();
+        loop {
+            let n = file.read(&mut buf[..])?;
+            if n == 0 {
+                break;
+            }
+            parser.add_input(&buf[..n])?;
+        }
+        parser.finish()
+    }
+
+    // Parse a SymbolFile from a file.
+    pub async fn from_file_async(path: &Path) -> Result<SymbolFile, SymbolError> {
+        use tokio::io::AsyncReadExt;
+
+        let mut file = tokio::fs::File::open(path).await?;
+        let mut buf = vec![0; 1024 * 8];
+        let mut parser = StreamingSymbolParser::new();
+        loop {
+            let n = file.read(&mut buf[..]).await?;
+            if n == 0 {
+                break;
+            }
+            parser.add_input(&buf[..n])?;
+        }
+        parser.finish()
```

Comment:
Here's a minimal example of driving the new streaming parser with a random async source.